File: mobile/ios/CommunityMaps/ViewModels/MapViewModel.swift
import Foundation
import MapKit
import CoreLocation

@MainActor
class MapViewModel: NSObject, ObservableObject, CLLocationManagerDelegate {
    @Published var alerts: [Alert] = []
    @Published var categories: [Category] = []
    @Published var activeFilters: Set<String> = []
    @Published var region = MKCoordinateRegion(
        center: CLLocationCoordinate2D(latitude: 40.7128, longitude: -74.006),
        span: MKCoordinateSpan(latitudeDelta: 0.05, longitudeDelta: 0.05)
    )
    @Published var isLoggedIn = false
    @Published var username: String?
    @Published var nearbyCount = 0
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let locationManager = CLLocationManager()
    private var userLocation: CLLocationCoordinate2D?

    var filteredAlerts: [Alert] {
        alerts.filter { activeFilters.contains($0.category) }
    }

    override init() {
        super.init()
        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
        checkAuth()
    }

    // MARK: - Data Loading
    func loadData() async {
        isLoading = true
        do {
            async let alertsTask = APIService.shared.getAlerts()
            async let categoriesTask = APIService.shared.getCategories()

            let (fetchedAlerts, fetchedCategories) = try await (alertsTask, categoriesTask)

            self.alerts = fetchedAlerts
            self.categories = fetchedCategories
            self.activeFilters = Set(fetchedCategories.map { $0.id })

            if let location = userLocation {
                await loadNearbyAlerts(lat: location.latitude, lng: location.longitude)
            }
        } catch {
            errorMessage = "Failed to load data: \(error.localizedDescription)"
        }
        isLoading = false
    }

    func loadNearbyAlerts(lat: Double, lng: Double) async {
        do {
            let nearby = try await APIService.shared.getNearbyAlerts(lat: lat, lng: lng, radius: 2)
            nearbyCount = nearby.count
        } catch {
            print("Failed to load nearby alerts: \(error)")
        }
    }

    func search(query: String) {
        Task {
            do {
                if query.isEmpty {
                    alerts = try await APIService.shared.getAlerts()
                } else {
                    alerts = try await APIService.shared.getAlerts(search: query)
                }
            } catch {
                errorMessage = "Search failed: \(error.localizedDescription)"
            }
        }
    }

    // MARK: - Filters
    func toggleFilter(_ categoryId: String) {
        if activeFilters.contains(categoryId) {
            activeFilters.remove(categoryId)
        } else {
            activeFilters.insert(categoryId)
        }
    }

    // MARK: - Alert Actions
    func createAlert(_ request: NewAlertRequest) async throws {
        let newAlert = try await APIService.shared.createAlert(request)
        alerts.insert(newAlert, at: 0)
    }

    func upvoteAlert(_ id: String) async {
        do {
            let updated = try await APIService.shared.upvoteAlert(id: id)
            if let index = alerts.firstIndex(where: { $0.id == id }) {
                alerts[index] = updated
            }
        } catch {
            errorMessage = "Failed to upvote: \(error.localizedDescription)"
        }
    }

    func deleteAlert(_ id: String) async {
        do {
            try await APIService.shared.deleteAlert(id: id)
            alerts.removeAll { $0.id == id }
        } catch {
            errorMessage = "Failed to delete: \(error.localizedDescription)"
        }
    }

    // MARK: - Location
    func requestLocationPermission() {
        locationManager.requestWhenInUseAuthorization()
    }

    func centerOnUser() {
        if let location = userLocation {
            region = MKCoordinateRegion(
                center: location,
                span: MKCoordinateSpan(latitudeDelta: 0.01, longitudeDelta: 0.01)
            )
        } else {
            requestLocationPermission()
            locationManager.startUpdatingLocation()
        }
    }

    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }
        userLocation = location.coordinate
        locationManager.stopUpdatingLocation()

        Task {
            await loadNearbyAlerts(lat: location.coordinate.latitude, lng: location.coordinate.longitude)
        }
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        print("Location error: \(error)")
    }

    // MARK: - Auth
    func checkAuth() {
        if let userData = UserDefaults.standard.data(forKey: "user"),
           let user = try? JSONDecoder().decode(User.self, from: userData) {
            isLoggedIn = true
            username = user.username
        }
    }

    func login(email: String, password: String) async throws {
        let response = try await APIService.shared.login(email: email, password: password)
        let userData = try JSONEncoder().encode(response.user)
        UserDefaults.standard.set(userData, forKey: "user")
        isLoggedIn = true
        username = response.user.username
    }

    func register(username: String, email: String, password: String) async throws {
        let response = try await APIService.shared.register(username: username, email: email, password: password)
        let userData = try JSONEncoder().encode(response.user)
        UserDefaults.standard.set(userData, forKey: "user")
        isLoggedIn = true
        self.username = response.user.username
    }

    func logout() {
        APIService.shared.logout()
        isLoggedIn = false
        username = nil
    }
}

File: mobile/android/app/src/main/java/com/communitymaps/viewmodels/MapViewModel.kt
package com.communitymaps.viewmodels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.communitymaps.models.*
import com.communitymaps.services.RetrofitClient
import com.google.android.gms.maps.model.LatLng
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class MapViewModel(application: Application) : AndroidViewModel(application) {

    private val apiService = RetrofitClient.apiService
    private val prefs = application.getSharedPreferences("auth", 0)

    private val _alerts = MutableStateFlow<List<Alert>>(emptyList())
    val alerts: StateFlow<List<Alert>> = _alerts.asStateFlow()

    private val _categories = MutableStateFlow<List<Category>>(emptyList())
    val categories: StateFlow<List<Category>> = _categories.asStateFlow()

    private val _activeFilters = MutableStateFlow<Set<String>>(emptySet())
    val activeFilters: StateFlow<Set<String>> = _activeFilters.asStateFlow()

    private val _isLoggedIn = MutableStateFlow(false)
    val isLoggedIn: StateFlow<Boolean> = _isLoggedIn.asStateFlow()

    private val _username = MutableStateFlow<String?>(null)
    val username: StateFlow<String?> = _username.asStateFlow()

    private val _nearbyCount = MutableStateFlow(0)
    val nearbyCount: StateFlow<Int> = _nearbyCount.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _mapCenter = MutableStateFlow(LatLng(40.7128, -74.006))
    val mapCenter: StateFlow<LatLng> = _mapCenter.asStateFlow()

    val filteredAlerts: List<Alert>
        get() = _alerts.value.filter { _activeFilters.value.contains(it.category) }

    private var token: String?
        get() = prefs.getString("token", null)
        set(value) = prefs.edit().putString("token", value).apply()

    init {
        checkAuth()
        loadData()
    }

    fun loadData() {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                val alertsResponse = apiService.getAlerts()
                val categoriesResponse = apiService.getCategories()

                if (alertsResponse.isSuccessful) {
                    _alerts.value = alertsResponse.body() ?: emptyList()
                }

                if (categoriesResponse.isSuccessful) {
                    val cats = categoriesResponse.body() ?: emptyList()
                    _categories.value = cats
                    _activeFilters.value = cats.map { it.id }.toSet()
                }
            } catch (e: Exception) {
                _errorMessage.value = "Failed to load data: ${e.message}"
            }
            _isLoading.value = false
        }
    }

    fun search(query: String) {
        viewModelScope.launch {
            try {
                val response = if (query.isEmpty()) {
                    apiService.getAlerts()
                } else {
                    apiService.getAlerts(search = query)
                }
                if (response.isSuccessful) {
                    _alerts.value = response.body() ?: emptyList()
                }
            } catch (e: Exception) {
                _errorMessage.value = "Search failed: ${e.message}"
            }
        }
    }

    fun toggleFilter(categoryId: String) {
        val current = _activeFilters.value.toMutableSet()
        if (current.contains(categoryId)) {
            current.remove(categoryId)
        } else {
            current.add(categoryId)
        }
        _activeFilters.value = current
    }

    fun loadNearbyAlerts(lat: Double, lng: Double) {
        viewModelScope.launch {
            try {
                val response = apiService.getNearbyAlerts(lat, lng, 2.0)
                if (response.isSuccessful) {
                    _nearbyCount.value = response.body()?.size ?: 0
                }
            } catch (e: Exception) {
                // Silently fail for nearby alerts
            }
        }
    }

    fun createAlert(request: NewAlertRequest) {
        viewModelScope.launch {
            try {
                val authToken = token?.let { "Bearer $it" } ?: ""
                val response = apiService.createAlert(authToken, request)
                if (response.isSuccessful) {
                    response.body()?.let { newAlert ->
                        _alerts.value = listOf(newAlert) + _alerts.value
                    }
                }
            } catch (e: Exception) {
                _errorMessage.value = "Failed to create alert: ${e.message}"
            }
        }
    }

    fun upvoteAlert(id: String) {
        viewModelScope.launch {
            try {
                val response = apiService.upvoteAlert(id)
                if (response.isSuccessful) {
                    response.body()?.let { updated ->
                        _alerts.value = _alerts.value.map {
                            if (it.id == id) updated else it
                        }
                    }
                }
            } catch (e: Exception) {
                _errorMessage.value = "Failed to upvote: ${e.message}"
            }
        }
    }

    fun deleteAlert(id: String) {
        viewModelScope.launch {
            try {
                val authToken = token?.let { "Bearer $it" } ?: ""
                val response = apiService.deleteAlert(authToken, id)
                if (response.isSuccessful) {
                    _alerts.value = _alerts.value.filter { it.id != id }
                }
            } catch (e: Exception) {
                _errorMessage.value = "Failed to delete: ${e.message}"
            }
        }
    }

    fun login(email: String, password: String, onSuccess: () -> Unit, onError: (String) -> Unit) {
        viewModelScope.launch {
            try {
                val response = apiService.login(LoginRequest(email, password))
                if (response.isSuccessful) {
                    response.body()?.let {
                        token = it.token
                        prefs.edit().putString("username", it.user.username).apply()
                        _isLoggedIn.value = true
                        _username.value = it.user.username
                        onSuccess()
                    }
                } else {
                    onError("Invalid credentials")
                }
            } catch (e: Exception) {
                onError(e.message ?: "Login failed")
            }
        }
    }

    fun register(username: String, email: String, password: String, onSuccess: () -> Unit, onError: (String) -> Unit) {
        viewModelScope.launch {
            try {
                val response = apiService.register(RegisterRequest(username, email, password))
                if (response.isSuccessful) {
                    response.body()?.let {
                        token = it.token
                        prefs.edit().putString("username", it.user.username).apply()
                        _isLoggedIn.value = true
                        _username.value = it.user.username
                        onSuccess()
                    }
                } else {
                    onError("Registration failed")
                }
            } catch (e: Exception) {
                onError(e.message ?: "Registration failed")
            }
        }
    }

    fun logout() {
        token = null
        prefs.edit().remove("username").apply()
        _isLoggedIn.value = false
        _username.value = null
    }

    private fun checkAuth() {
        if (token != null) {
            _isLoggedIn.value = true
            _username.value = prefs.getString("username", null)
        }
    }

    fun setMapCenter(latLng: LatLng) {
        _mapCenter.value = latLng
    }
}
